### PR TITLE
Use getMaybeExpandedExtent() for lowering and launching segments

### LINKED ISSUE #469: Failure in prepareRuntimeOrder for possibly expanded input
Some of the tests in #434 are failing in `FusionKernelRuntime::prepareRuntimeOrder` (pull that branch then `pytest python_tests/test_normalization.py::test_match[float32-False-False-True-contiguous-5-1-3-True]`). Here is the python repro:
```python
import torch
from nvfuser import FusionDefinition, DataType

def nvfuser_fusion_id0(fd : FusionDefinition) -> None :
    T0 = fd.define_tensor(symbolic_sizes=[-1, -1, -1, -1, -1], contiguity=[None, True, True, True, True], dtype=DataType.Float, is_cpu=False)
    T1 = fd.define_tensor(symbolic_sizes=[-1], contiguity=[True], dtype=DataType.Float, is_cpu=False)
    T2 = fd.define_tensor(symbolic_sizes=[-1], contiguity=[True], dtype=DataType.Float, is_cpu=False)
    S3 = fd.define_scalar(None, dtype=DataType.Double)
    S4 = fd.define_scalar(None, dtype=DataType.Double)
    S5, S6, S7, S8, S9, = fd.ops.tensor_sizes(T0)
    S10 = fd.define_scalar(1, dtype=DataType.Int)
    S11 = fd.ops.mul(S10, S5)
    S12 = fd.ops.mul(S11, S9)
    S13 = fd.define_scalar(1, dtype=DataType.Int)
    S14 = fd.ops.mul(S13, S6)
    S15 = fd.ops.mul(S14, S7)
    S16 = fd.ops.mul(S15, S8)
    T17, T18 = fd.ops.var_mean(T0, axes=[1, 2, 3], correction=0, keepdim=False)
    S19 = fd.define_scalar(1.00000, dtype=DataType.Double)
    S20 = fd.ops.sub(S19, S3)
    T21 = fd.ops.mul(T18, S3)
    T22 = fd.ops.mul(T1, S20)
    T23 = fd.ops.add(T22, T21)
    T24 = fd.ops.sum(T23, axes=[0], keepdim=False, dtype=DataType.Null)
    S25 = fd.ops.reciprocal(S5)
    T26 = fd.ops.mul(T24, S25)
    fd.add_output(T26, T1)
    S27 = fd.define_scalar(1, dtype=DataType.Int)
    S28 = fd.ops.sub(S16, S27)
    S29 = fd.ops.div(S16, S28)
    T30 = fd.ops.mul(T17, S29)
    T31 = fd.ops.mul(T30, S3)
    T32 = fd.ops.mul(T2, S20)
    T33 = fd.ops.add(T32, T31)
    T34 = fd.ops.sum(T33, axes=[0], keepdim=False, dtype=DataType.Null)
    T35 = fd.ops.mul(T34, S25)
    fd.add_output(T35, T2)
    T36 = fd.ops.broadcast_in_dim(T18, output_shape=[S5, S6, S7, S8, S9], broadcast_dims=[0, 4])
    T37 = fd.ops.sub(T0, T36)
    T38 = fd.ops.add(T17, S4)
    T39 = fd.ops.rsqrt(T38)
    T40 = fd.ops.broadcast_in_dim(T39, output_shape=[S5, S6, S7, S8, S9], broadcast_dims=[0, 4])
    T41 = fd.ops.mul(T37, T40)
    fd.add_output(T41)
    fd.add_output(T18)
    fd.add_output(T39)

with FusionDefinition() as fd:
    nvfuser_fusion_id0(fd)

inputs = [
    torch.randn((1, 3, 3, 3, 7), dtype=torch.float32, device='cuda:0').as_strided((1, 3, 3, 3, 7), (189, 63, 21, 7, 1)),
    torch.randn((7,), dtype=torch.float32, device='cuda:0').as_strided((7,), (1,)),
    torch.randn((7,), dtype=torch.float32, device='cuda:0').as_strided((7,), (1,)),
    0.1,
    1e-05,
]
fd.execute(inputs)
```
> RuntimeError: one_ran INTERNAL ASSERT FAILED at "/opt/pytorch/nvfuser/csrc/kernel_cache.cpp":814, please report a bug to PyTorch. Couldn't run all groups, something must have gone wrong in segmentation.

Notice that the input has size 1 in the batch dimension, but when `T0` is defined it has `None` contiguity but `-1` symbolic size. That's because this code is a relic from when `fd.from_pytorch` did not exist; however I think this is how we would expect to receive expanded inputs. Changing the first `-1` to `1` in `T0`'s `symbolic_sizes` is a workaround. However, I think it should either give an informative error in these cases or run properly.

This `Fusion` is segmented into three segments, and 
```
Available inputs:
  d10
  i7
  d9
  T1_g[ iS5{i7} ]
  i8
  T2_g[ iS6{i8} ]
  i4
  i3
  i2
  i1
  1
  T0_g[ bS0{i0}, iS1{i1}, iS2{i2}, iS3{i3}, iS4{i4} ]
group 0 inputs:
  T1_g[ iS5{i7} ]
  d9
  T3_g[ bS7{i0}, rS8{i1}, rS9{i2}, rS10{i3}, iS11{i4} ]
  i0
group 0 outputs:
  T12_g[ iS33{i7} ]
group 1 inputs:
  T2_g[ iS6{i8} ]
  d9
  T13_g[ bS34{i0}, iS35{i4} ]
  i0
group 1 outputs:
  T19_g[ iS45{i8} ]
group 2 inputs:
  d10
  T0_g[ bS0{i0}, iS1{i1}, iS2{i2}, iS3{i3}, iS4{i4} ]
  i0
  i1
  i2
  i3
  i4
group 2 outputs:
  T24_g[ bS63{i0}, iS64{i4} ]
  T3_g[ bS7{i0}, rS8{i1}, rS9{i2}, rS10{i3}, iS11{i4} ]
  T27_g[ bS75{i0}, iS76{i1}, iS77{i2}, iS78{i3}, iS79{i4} ]
```
These groups should be launched in order 2, 0, 1. However, group 2 can't be launched because the first extent `i0` is not available. This is because `bS0{i0}` has extent `1` but its `expandedExtent` is `i0`, and we only make `extent()` available as an input. In order to properly accept expanded inputs, we should probably at minimum change `FusionKernelRuntime::prepareRuntimeOrder` to use `getMaybeExpandedExtent()` instead of `extent()` when finding available inputs. That happens to fix this particular issue, but more testing should probably done specifically for expanded inputs in segmented Fusions.

### LINKED ISSUE #470: Compilation failure for segment input with pre-simplified expression as expanded extent
Running `python_tests/test_normalization.py::test_match[float32-True-False-True-contiguous-1-7-3-True]` on #434.

Python repro:
```python
import torch
from nvfuser import FusionDefinition, DataType

def nvfuser_fusion_id51(fd : FusionDefinition) -> None :
    T0 = fd.define_tensor(symbolic_sizes=[-1, -1, -1, -1, -1], contiguity=[None, True, True, True, True], dtype=DataType.Float, is_cpu=False)
    T1 = fd.define_tensor(symbolic_sizes=[-1], contiguity=[True], dtype=DataType.Float, is_cpu=False)
    T2 = fd.define_tensor(symbolic_sizes=[-1], contiguity=[True], dtype=DataType.Float, is_cpu=False)
    T3 = fd.define_tensor(symbolic_sizes=[-1, -1], contiguity=[None, True], dtype=DataType.Float, is_cpu=False)
    T4 = fd.define_tensor(symbolic_sizes=[-1, -1], contiguity=[None, True], dtype=DataType.Float, is_cpu=False)
    T5 = fd.define_tensor(symbolic_sizes=[-1, -1, -1, -1, -1], contiguity=[None, True, True, True, True], dtype=DataType.Float, is_cpu=False)
    S6, S7, S8, S9, S10, = fd.ops.tensor_sizes(T0)
    S11 = fd.define_scalar(1, dtype=DataType.Int)
    S12 = fd.ops.mul(S11, S6)
    S13 = fd.ops.mul(S12, S7)
    S14 = fd.define_scalar(1, dtype=DataType.Int)
    S15 = fd.ops.mul(S14, S8)
    S16 = fd.ops.mul(S15, S9)
    S17 = fd.ops.mul(S16, S10)
    T18 = fd.ops.broadcast_in_dim(T3, output_shape=[S6, S7, S8, S9, S10], broadcast_dims=[0, 1])
    S19 = fd.ops.reciprocal(S17)
    T20 = fd.ops.sum(T5, axes=[2, 3, 4], keepdim=False, dtype=DataType.Null)
    T21 = fd.ops.sub(T0, T18)
    T22 = fd.ops.mul(T5, T21)
    T23 = fd.ops.sum(T22, axes=[2, 3, 4], keepdim=False, dtype=DataType.Null)
    T24 = fd.ops.mul(T20, S19)
    T25 = fd.ops.broadcast_in_dim(T24, output_shape=[S6, S7, S8, S9, S10], broadcast_dims=[0, 1])
    T26 = fd.ops.mul(T23, S19)
    T27 = fd.ops.mul(T4, T4)
    T28 = fd.ops.mul(T26, T27)
    T29 = fd.ops.broadcast_in_dim(T28, output_shape=[S6, S7, S8, S9, S10], broadcast_dims=[0, 1])
    T30 = fd.ops.broadcast_in_dim(T4, output_shape=[S6, S7, S8, S9, S10], broadcast_dims=[0, 1])
    T31 = fd.ops.broadcast_in_dim(T1, output_shape=[S6, S7, S8, S9, S10], broadcast_dims=[1])
    T32 = fd.ops.mul(T30, T31)
    T33 = fd.ops.sub(T0, T18)
    T34 = fd.ops.mul(T33, T29)
    T35 = fd.ops.sub(T5, T34)
    T36 = fd.ops.sub(T35, T25)
    T37 = fd.ops.mul(T36, T32)
    T38 = fd.ops.mul(T23, T4)
    T39 = fd.ops.sum(T38, axes=[0], keepdim=False, dtype=DataType.Null)
    T40 = fd.ops.sum(T20, axes=[0], keepdim=False, dtype=DataType.Null)
    fd.add_output(T37)
    fd.add_output(T39)
    fd.add_output(T40)

with FusionDefinition() as fd:
    nvfuser_fusion_id51(fd)

inputs = [
    torch.randn((1, 7, 3, 3, 3), dtype=torch.float32, device='cuda:0').as_strided((1, 7, 3, 3, 3), (189, 27, 9, 3, 1)),
    torch.randn((7,), dtype=torch.float32, device='cuda:0').as_strided((7,), (1,)),
    torch.randn((7,), dtype=torch.float32, device='cuda:0').as_strided((7,), (1,)),
    torch.randn((1, 7), dtype=torch.float32, device='cuda:0').as_strided((1, 7), (0, 1)),
    torch.randn((1, 7), dtype=torch.float32, device='cuda:0').as_strided((1, 7), (0, 1)),
    torch.randn((1, 7, 3, 3, 3), dtype=torch.float32, device='cuda:0').as_strided((1, 7, 3, 3, 3), (189, 27, 9, 3, 1)),
]
fd.execute(inputs)
```
This actually creates a kernel that fails to compile:
```c++
__global__ void kernel2(Tensor<float, 2, 2> T4, Tensor<float, 2, 2> T11, Tensor<float, 1, 1> T31) {
  int i0;
  i0 = ((nvfuser_index_t)threadIdx.x) + (((nvfuser_index_t)blockDim.x) * ((nvfuser_index_t)blockIdx.x));
  bool b1;
  b1 = i0 < T4.size[1];
  float T30[1];
  if (b1) {
    T30[0]
      = T11[i0]
      * T4[i0];
  }
  float T33[1];
  T33[0] = 0.00000000000000000e+00;
  #pragma unroll 1
  for(nvfuser_index_t i2 = 0; i2 < i3; ++i2) {
    if (b1) {
      T33[0]
        = T33[0]
        + T30[0];
    }
  }
  if (b1) {
    T31[i0] = 0.00000000000000000e+00;
  }
  if (b1) {
    T31[i0]
      = T31[i0]
      + T33[0];
  }
}
}
/*
CUDA NVRTC compile error: __tmp_kernel2.cu(9564): error: identifier "i3" is undefined
    for(nvfuser_index_t i2 = 0; i2 < i3; ++i2) {
                                     ^

1 error detected in the compilation of "__tmp_kernel2.cu".
*/
```
The fusion math for this segment is
```
Inputs:
  T4_g[ iS148{( ceilDiv(i12, blockDim.x) )}, iS149{blockDim.x}, bS150{( ceilDiv(i11, 1) )}, bS151{1} ], float
  T11_g[ iS152{( ceilDiv(i14, blockDim.x) )}, iS153{blockDim.x}, bS154{( ceilDiv(i13, 1) )}, bS155{1}, rS43{i15}, rS44{i16}, rS45{i17} ], float
Outputs:
  T31_g[ iblockIdx.x142{( ceilDiv(i14, blockDim.x) )}, ithreadIdx.x143{blockDim.x}, rUS140{1} ] ca_pos( 2 ) produce_pos( 2 ), float

%kernel_math {
T30_l[ iblockIdx.x144{( ceilDiv(i14, blockDim.x) )}, ithreadIdx.x145{blockDim.x}, bS146{( ceilDiv(i13, 1) )}, bUS147{1} ] ca_pos( 2 )
   = T11_g[ iS152{( ceilDiv(i14, blockDim.x) )}, iS153{blockDim.x}, bS154{( ceilDiv(i13, 1) )}, bS155{1}, rS43{i15}, rS44{i16}, rS45{i17} ]
   * T4_g[ iS148{( ceilDiv(i12, blockDim.x) )}, iS149{blockDim.x}, bS150{( ceilDiv(i11, 1) )}, bS151{1} ];
T33_l[ iblockIdx.x136{( ceilDiv(i14, blockDim.x) )}, ithreadIdx.x137{blockDim.x}, rS138{( ceilDiv(i13, 1) )}rf, iUS139{1}rf ] ca_pos( 2 ) produce_pos( 2 )
   = reduction( T30_l[ iblockIdx.x144{( ceilDiv(i14, blockDim.x) )}, ithreadIdx.x145{blockDim.x}, bS146{( ceilDiv(i13, 1) )}, bUS147{1} ] ca_pos( 2 ), op = add, initial value = double(0), allreduce = false )
T31_g[ iblockIdx.x142{( ceilDiv(i14, blockDim.x) )}, ithreadIdx.x143{blockDim.x}, rUS140{1} ] ca_pos( 2 ) produce_pos( 2 )
   = reduction( T33_l[ iblockIdx.x136{( ceilDiv(i14, blockDim.x) )}, ithreadIdx.x137{blockDim.x}, rS138{( ceilDiv(i13, 1) )}rf, iUS139{1}rf ] ca_pos( 2 ) produce_pos( 2 ), op = add, initial value = double(0), allreduce = false )
}
```
We see that `i13` appears in an expanded extent in `T11_g`. The actual extent is `ceilDiv(i13, 1)` which probably gets simplified to `i13`, and although `ceilDiv(i13, 1)` is bound, `i13` is not. This is a case where it might be nice to have a union-find so that we don't lose track of computable scalars (see #382).

### PR DESCRIPTION
**Pull request by jacobhinkle**
Previously some of the lowering and launching code implicitly assumed inputs were not expanded broadcasts. This PR addresses those.

Fixes #469 and fixes #470.

### CODE REVIEW

**Comment by jacobhinkle** [comment]
I'm currently hitting:
```
>       self.assertEqual(fc.num_fusions() - before_fusions, int(new_fusion_expected))
E       AssertionError: Scalars are not equal!
E       
E       Expected 0 but got 1.
E       Absolute difference: 1
E       Relative difference: inf

python_tests/test_python_frontend.py:114: AssertionError
```
It seems that the deserialized fusion differs from the re-executed one. Using `PYTORCH_NVFUSER_DUMP=python_definition pytest python_tests/test_python_frontend.py::TestNvFuserFrontend::test_repro469` it appears the printout of the two fusions are identical (except for the id number):
```python
def nvfuser_fusion_id0(fd : FusionDefinition) -> None :
    T0 = fd.define_tensor(symbolic_sizes=[-1, -1, -1, -1, -1], contiguity=[None, True, True, True, True], dtype=DataType.Float, is_cpu=False)
    T1 = fd.define_tensor(symbolic_sizes=[-1], contiguity=[True], dtype=DataType.Float, is_cpu=False)
    T2 = fd.define_tensor(symbolic_sizes=[-1], contiguity=[True], dtype=DataType.Float, is_cpu=False)
    S3 = fd.define_scalar(None, dtype=DataType.Double)
    S4 = fd.define_scalar(None, dtype=DataType.Double)
    S5, S6, S7, S8, S9, = fd.ops.tensor_sizes(T0)
    S10 = fd.define_scalar(1, dtype=DataType.Int)
    S11 = fd.ops.mul(S10, S5)
    S12 = fd.ops.mul(S11, S9)
    S13 = fd.define_scalar(1, dtype=DataType.Int)
    S14 = fd.ops.mul(S13, S6)
    S15 = fd.ops.mul(S14, S7)
    S16 = fd.ops.mul(S15, S8)
    T17, T18 = fd.ops.var_mean(T0, axes=[1, 2, 3], correction=0, keepdim=False)
    S19 = fd.define_scalar(1.00000, dtype=DataType.Double)
    S20 = fd.ops.sub(S19, S3)
    T21 = fd.ops.mul(T18, S3)
    T22 = fd.ops.mul(T1, S20)
    T23 = fd.ops.add(T22, T21)
    T24 = fd.ops.sum(T23, axes=[0], keepdim=False, dtype=DataType.Null)
    S25 = fd.ops.reciprocal(S5)
    T26 = fd.ops.mul(T24, S25)
    fd.add_output(T26, T1)
    S27 = fd.define_scalar(1, dtype=DataType.Int)
    S28 = fd.ops.sub(S16, S27)
    S29 = fd.ops.div(S16, S28)
    T30 = fd.ops.mul(T17, S29)
    T31 = fd.ops.mul(T30, S3)
    T32 = fd.ops.mul(T2, S20)
    T33 = fd.ops.add(T32, T31)
    T34 = fd.ops.sum(T33, axes=[0], keepdim=False, dtype=DataType.Null)
    T35 = fd.ops.mul(T34, S25)
    fd.add_output(T35, T2)
    T36 = fd.ops.broadcast_in_dim(T18, output_shape=[S5, S6, S7, S8, S9], broadcast_dims=[0, 4])
    T37 = fd.ops.sub(T0, T36)
    T38 = fd.ops.add(T17, S4)
    T39 = fd.ops.rsqrt(T38)
    T40 = fd.ops.broadcast_in_dim(T39, output_shape=[S5, S6, S7, S8, S9], broadcast_dims=[0, 4])
    T41 = fd.ops.mul(T37, T40)
    fd.add_output(T41)
    fd.add_output(T18)
    fd.add_output(T39)

def nvfuser_fusion_id1(fd : FusionDefinition) -> None :
    T0 = fd.define_tensor(symbolic_sizes=[-1, -1, -1, -1, -1], contiguity=[None, True, True, True, True], dtype=DataType.Float, is_cpu=False)
    T1 = fd.define_tensor(symbolic_sizes=[-1], contiguity=[True], dtype=DataType.Float, is_cpu=False)
    T2 = fd.define_tensor(symbolic_sizes=[-1], contiguity=[True], dtype=DataType.Float, is_cpu=False)
    S3 = fd.define_scalar(None, dtype=DataType.Double)
    S4 = fd.define_scalar(None, dtype=DataType.Double)
    S5, S6, S7, S8, S9, = fd.ops.tensor_sizes(T0)
    S10 = fd.define_scalar(1, dtype=DataType.Int)
    S11 = fd.ops.mul(S10, S5)
    S12 = fd.ops.mul(S11, S9)
    S13 = fd.define_scalar(1, dtype=DataType.Int)
    S14 = fd.ops.mul(S13, S6)
    S15 = fd.ops.mul(S14, S7)
    S16 = fd.ops.mul(S15, S8)
    T17, T18 = fd.ops.var_mean(T0, axes=[1, 2, 3], correction=0, keepdim=False)
    S19 = fd.define_scalar(1.00000, dtype=DataType.Double)
    S20 = fd.ops.sub(S19, S3)
    T21 = fd.ops.mul(T18, S3)
    T22 = fd.ops.mul(T1, S20)
    T23 = fd.ops.add(T22, T21)
    T24 = fd.ops.sum(T23, axes=[0], keepdim=False, dtype=DataType.Null)
    S25 = fd.ops.reciprocal(S5)
    T26 = fd.ops.mul(T24, S25)
    fd.add_output(T26, T1)
    S27 = fd.define_scalar(1, dtype=DataType.Int)
    S28 = fd.ops.sub(S16, S27)
    S29 = fd.ops.div(S16, S28)
    T30 = fd.ops.mul(T17, S29)
    T31 = fd.ops.mul(T30, S3)
    T32 = fd.ops.mul(T2, S20)
    T33 = fd.ops.add(T32, T31)
    T34 = fd.ops.sum(T33, axes=[0], keepdim=False, dtype=DataType.Null)
    T35 = fd.ops.mul(T34, S25)
    fd.add_output(T35, T2)
    T36 = fd.ops.broadcast_in_dim(T18, output_shape=[S5, S6, S7, S8, S9], broadcast_dims=[0, 4])
    T37 = fd.ops.sub(T0, T36)
    T38 = fd.ops.add(T17, S4)
    T39 = fd.ops.rsqrt(T38)
    T40 = fd.ops.broadcast_in_dim(T39, output_shape=[S5, S6, S7, S8, S9], broadcast_dims=[0, 4])
    T41 = fd.ops.mul(T37, T40)
    fd.add_output(T41)
    fd.add_output(T18)
    fd.add_output(T39)
```

File: python_tests/test_python_frontend.py [hunk 1]
```diff
@@ -62,19 +62,16 @@ def inner(*args, **kwargs):
...
+            FusionCache.reset()
 
-        FusionCache.reset()
-
-        # Get new FusionCache because the previous one was destroyed by the reset call.
-        fc = FusionCache.get()
-        fc.deserialize("foo.bin")
+            # Get new FusionCache because the previous one was destroyed by the reset call.
+            fc = FusionCache.get()
+            fc.deserialize(tmp.name)
```
Comment (jacobhinkle) [inline python_tests/test_python_frontend.py:74]: Just cleaning up the `foo.bin` stuff while I'm here. Reverting that commit has no effect on the error I'm seeing.

**Comment by rdspring1** [comment]
I'm looking at `PYTORCH_NVFUSER_DUMP=python_frontend_debug` to see why the fusion cache thinks they're different.

File: python_tests/test_python_frontend.py [hunk 2]
```diff
@@ -2258,6 +2255,95 @@ def fusion_func(fd: FusionDefinition):
...
+                dtype=DataType.Float,
+                is_cpu=False,
+            )
+            T2 = fd.define_tensor(
+                symbolic_sizes=[-1],
+                contiguity=[True],
+                dtype=DataType.Float,
+                is_cpu=False,
+            )
+            S3 = fd.define_scalar(None, dtype=DataType.Double)
```
Comment (rdspring1) [inline python_tests/test_python_frontend.py:2280]:
```suggestion
            S3 = fd.define_scalar(dtype=DataType.Double)
```
This ^^^ works for symbolic scalar inputs.
For some reason with the optional argument, the `ValueType` for ScalarRecord becomes `bool`.

Comment (jacobhinkle) [reply python_tests/test_python_frontend.py:2280]: Yeah that is strange. In this case, the test comes from the print-out of a python exception. So one temporary workaround would be to not print `None, ` in `ScalarRecord::print()` for input scalars. The heart of the issue though seems to be that `ScalarRecord::operator==()` does not behave properly when `value_ == std::nullopt`.

**Comment by jacobhinkle** [comment]
!build

**Comment by jacobhinkle** [comment]
> Are @rdspring1's concerns resolved? @jacobhinkle, @rdspring1

Ryan can confirm but I think so as of #484.

**Comment by rdspring1** [comment]
Yes, https://github.com/NVIDIA/Fuser/pull/484 fixed the problem.

**Review by zasdfgbnm — APPROVED** [review]
I in the past added a test for the case where expanded tensor can be fusion inputs in https://github.com/csarofeen/pytorch/pull/2049
Looks like that test is not catching all required changes.

**Pull request merged**